lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -658,11 +668,15 @@ def _handle_endpoints_changed(self, event: RelationEvent) -> None:
     def _push_alerts_on_relation_joined(self, event: RelationEvent) -> None:
         self._push_alerts_to_relation_databag(event.relation)
 
-    def _push_alerts_to_all_relation_databags(self, _: Optional[HookEvent]) -> None:
+    def _push_alerts_to_all_relation_databags(
+        self, _: Optional[HookEvent], *, extra_alerts_list: Optional[list] = None
+    ) -> None:
```

Comment:
I mean, in general, I think this whole patch is ok.

Conceptually, though, `_push_alerts_to_relation_databags` also triggers on `leader-elected`, `upgrade-charm`, and `relation-joined`. These do not trigger `src.PrometheusCharm._push_alerts_to_remote_write`.

It seems entirely possible for some other event (a new leader election after scaling the application, refreshing the charm, joining a new consumer)  to then invoke `_push_alerts_to_relation_databag` _without_ anything in `extra_alerts_list`, and they will silently disappear. I would suspect this could even be verified in `test_receive_remote_write` with an additional testcase which performed one of these (refreshing the charm, for example, since there are... problem with scaling in integration tests right now).

We really don't always love `StoredState`, or forcing the existence of a peer relation, but those are about the only two good mechanisms of persistence between events with dynamic data this way.

Alternatively, we did something for Kubeflow endpoint discovery where `extra_alerts_list` could be something like `lookaside_alerts_callable`, set in the constructor, and that callable (which could as easily be `PrometheusCharm.metrics_consumer.alerts`) is invoked if it exists and is callable, which provides kind of an abstraction which doesn't _require_ users to break through encapsulation by directly twiddling `metrics_consumer` from here, but presents the option to do so. Or nothing. Or a lambda. Or whatever.